Normative: Add Optional Chaining

DanielRosenwasser opened the pull request:
This change adds the relevant updates for the stage 3 [Optional Chaining](https://github.com/tc39/proposal-optional-chaining/) proposal. The intended changes can be viewed at https://tc39.es/proposal-optional-chaining/.

The Test262 PR is available at https://github.com/tc39/test262/pull/2212

ljharb commented:
@DanielRosenwasser please either link to the test262 tests PR, or put one up, prior to the December meeting. Thanks!

bakkot commented on spec.html:
```diff
@@ -13799,6 +13892,74 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
+          1. If _baseValue_ is *undefined* or *null*, then
+            1. Return *undefined*.
+          1. Let _optionalChain_ be this |OptionalChain|.
+          1. Return ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
```
I think this is less clear than just having three different `Runtime Semantics: Evaluation ` sections, one for each of the three kinds of chain, each with their own copy of the steps preceding this line.

DanielRosenwasser replied:
Because I spent a lot of time misunderstanding this, I should probably clarify before I do any more work 😄

Basically the line that you actually take issue with is this, right?

```
1. Let _baseExpression_ be the first child of this production (i.e., this |MemberExpression|, |CallExpression|, or |OptionalExpression|).
```

It feels repetitive to have 3 different sections, but I can make the change.

bakkot replied:
I do also take some issue with that line, but actually it's the method-style invocation of `ChainEvaluation` on the commented line which I find the least clear. The spec does allow for such invocations in some circumstances, but they're used almost exclusively for working with environment records, not to dispatch abstract operations on parse nodes.

I agree it is a tiny bit repetitive, but I think the added clarity is worth it. And it's not actually that much code which gets repeated.

devsnek replied:
I also found this to be a bit weird. +1 to separate runtime semantics.

zenparsing replied:
I agree with @bakkot here. In general, I think it's best to stick to the "syntax-directed evaluation" style as much as possible.

bakkot replied:
To be explicit, what I think would be the most clear would be to have three subsections under the current `Runtime Semantics: Evaluation`, each in the mold of

```
        <emu-grammar>
          OptionalExpression :
            MemberExpression OptionalChain
        </emu-grammar>
        <emu-alg>
          1. Let _baseReference_ be the result of evaluating |MemberExpression|.
          1. Let _baseValue_ be ? GetValue(_baseReference_).
          1. If _baseValue_ is *undefined* or *null*, then
            1. Return *undefined*.
          1. Return the result of performing ChainEvaluation for |OptionalChain| with arguments _baseValue_ and _baseReference_.
        <emu-alg>
```

(And to similarly rewrite the uses of `ChainEvaluation` in the `Runtime Semantics: ChainEvaluation` section to use this style of dispatch.)

jmdyck replied:
> it's the method-style invocation of `ChainEvaluation` on the commented line which I find the least clear. The spec does allow for such invocations in some circumstances, but they're used almost exclusively for working with environment records,

And module records, and internal methods.

> not to dispatch abstract operations on parse nodes.

But see PR #955.

claudepache replied:
> it's the method-style invocation of `ChainEvaluation` on the commented line which I find the least clear. The spec does allow for such invocations in some circumstances, but they're used almost exclusively for working with environment records,

That was reverted to the traditional form “ChainEvaluation of ... with arguments ...” in tc39/proposal-optional-chaining#118, but te change has not been forwarded to this PR.

bakkot commented on spec.html:
```diff
@@ -13799,6 +13892,74 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
+        </emu-grammar>
+        <emu-alg>
+          1. Let _baseExpression_ be the first child of this production (i.e., this |MemberExpression|, |CallExpression|, or |OptionalExpression|).
+          1. Let _baseReference_ be ? _baseExpression_.Evaluation().
```
```suggestion
          1. Let _baseReference_ be the result of evaluating _baseExpression_.
```

(There's no need for the `?`; that's [handled by](https://tc39.es/ecma262/#sec-getvalue) the `GetValue` in subsequent step.)

bakkot commented on spec.html:
```diff
@@ -13799,6 +13892,74 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
+        </emu-alg>
+        <emu-grammar>OptionalChain : OptionalChain `[` Expression `]`</emu-grammar>
+        <emu-alg>
+          1. Let _optionalChain_ be this |OptionalChain|.
```
```suggestion
          1. Let _optionalChain_ be |OptionalChain|.
```

The `this |Nonterminal|` wording refers to the the entire production, rather than one of the nonterminals on the RHS. Saying `Let _optionalChain_ be |OptionalChain|` would refer to the one on the RHS. See for example [the Evaluation semantics for](https://tc39.es/ecma262/#sec-function-calls-runtime-semantics-evaluation) `CallExpression : CallExpression Arguments`, which has exactly the same ambiguity, and which in step 1 uses a bare `CallExpression` to mean the one on the RHS and in step 3 uses `this CallExpression` to mean the full production.

Here this line only makes sense if it is intended to refer to the one one the RHS rather than the full production.

(Ditto most other uses of `this |NonTerminal|` in this PR.)

rkirsling replied:
This was fixed in https://github.com/tc39/proposal-optional-chaining/pull/120 but must have not been propagated to this PR.

bakkot commented on spec.html:
```diff
@@ -13799,6 +13892,74 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
+        <emu-alg>
+          1. Let _optionalChain_ be this |OptionalChain|.
+          1. Let _newReference_ be ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
+          1. Let _newValue_ be ? GetValue(_newReference_).
```
Should there not be a

```
          1. If _newValue_ is *undefined* or *null*, then
            1. Return *undefined*.
```
after this line? (Ditto the ChainEvaluation sections for ``OptionalChain `[` Expression `]`` and ``OptionalChain : OptionalChain Arguments``.)

rkirsling replied:
No, if we have `a?.b.c`, then the chain is `?.b.c`—_whether_ we evaluate a chain is conditional but the evaluation of the chain is not.

bakkot replied:
Oh, thank you, yes. If a `return` were necessary it would have occurred before this algorithm were reached.

syg commented:
Could someone update this PR to reflect the upstream fixes and editorial changes?

syg commented on spec.html:
```diff
@@ -13569,29 +13639,52 @@ <h1>Runtime Semantics: Evaluation</h1>
     </emu-clause>
 
+    <emu-clause id="sec-evaluate-dynamic-property-access" aoid="EvaluateDynamicPropertyAccess">
+      <h1>Runtime Semantics: EvaluateDynamicPropertyAccess(_baseValue_, _expression_, _strict_ )</h1>
```
Hm... I'm not a fan of the Static/Dynamic naming here. Maybe ExpressionPropertyAccess and IdentifierPropertyAccess? @bakkot @ljharb wdyt?

michaelficarra replied:
"static" and "computed" are the terms I typically see/use

syg replied:
I'm fine with that too, though I still prefer the more literal names -- doesn't leave room for misinterpreting what exactly is static or computed.

DanielRosenwasser replied:
What about `EvaluateExpressionKeyPropertyAccess` and `EvaluateIdentifierKeyPropertyAccess`?

michaelficarra replied:
I don't like `Identifier` in the name at all. It's an inline string-named property access. And computed property accesses are property accesses where the property name is computed by evaluating a given expression. So `ExpressionPropertyAccess` / `ComputedPropertyAccess` and `StringPropertyAccess` / `StaticPropertyAccess` / `FixedPropertyAccess` are best IMO.

I also don't like `Key` in the name unless it's preceded by `Property`.

claudepache replied:
Another option would be `EvaluateDotPropertyAccess` and `EvaluateBracketPropertyAccess` (although I’m not exactly a fan of that).

Note that, with the advent of private fields and methods, there may be a future `EvaluatePrivatePropertyAccess`.

DanielRosenwasser replied:
> Note that, with the advent of private fields and methods, there may be a future `EvaluatePrivatePropertyAccess`.

In which case I think that `Identifier` and `Expression` are probably the best differentiators - because then it can simply be named `EvaluatePrivateNamePropertyAccess` or `EvaluatePrivateNameKeyPropertyAccess`

DanielRosenwasser replied:
I've gone with the naming I mentioned above, since it feels like a good way to leave the door open to *PrivateName` keys, but I didn't use the `PropertyKey` naming scheme because `PropertyKeyPropertyAccess` gets a bit unwieldy. Long term, if there's an issue, I figure it can always be renamed.

DanielRosenwasser commented:
I've integrated the upstream changes and addressed feedback provided apart from the `Evaluate_______PropertyAccess` naming discussion which I'll try to address shortly

bakkot commented on spec.html:
```diff
@@ -13379,9 +13387,25 @@ <h2>Syntax</h2>
+        OptionalChain[?Yield, ?Await]  `[` Expression[+In, ?Yield, ?Await] `]`
+        OptionalChain[?Yield, ?Await] `.` IdentifierName
+        OptionalChain[?Yield, ?Await] Arguments[?Yield, ?Await]
+        OptionalChain[?Yield, ?Await] TemplateLiteral[?Yield, ?Await, +Tagged]
```
I would be happier if these were in the same order as they are in [CallExpression](https://tc39.es/ecma262/#prod-CallExpression), i.e.

```
      OptionalChain[Yield, Await] :
        `?.` Arguments[?Yield, ?Await]
        `?.` `[` Expression[+In, ?Yield, ?Await] `]`
        `?.` IdentifierName
        `?.` TemplateLiteral[?Yield, ?Await, +Tagged]
        OptionalChain[?Yield, ?Await] Arguments[?Yield, ?Await]
        OptionalChain[?Yield, ?Await]  `[` Expression[+In, ?Yield, ?Await] `]`
        OptionalChain[?Yield, ?Await] `.` IdentifierName
        OptionalChain[?Yield, ?Await] TemplateLiteral[?Yield, ?Await, +Tagged]
```

bakkot commented on spec.html:
```diff
@@ -13393,6 +13417,33 @@ <h2>Supplemental Syntax</h2>
+          </li>
+        </ul>
+        <emu-note>
+          <p>This production exists in order to prevent automatic semicolon insertion rules (<emu-xref href="#sec-automatic-semicolon-insertion"></emu-xref>) to be applied to the following code:</p>
```
```suggestion
          <p>This production exists in order to prevent automatic semicolon insertion rules (<emu-xref href="#sec-automatic-semicolon-insertion"></emu-xref>) from applying to the following code:</p>
```

bakkot commented on spec.html:
```diff
@@ -13814,6 +13907,98 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
+          1. If _baseValue_ is *undefined* or *null*, then
+            1. Return *undefined*.
+          1. Let _optionalChain_ be |OptionalChain|.
+          1. Return ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
```
```suggestion
          1. Return the result of performing ChainEvaluation of _optionalChain_ with arguments _baseValue_ and _baseReference_.
```

bakkot commented on spec.html:
```diff
@@ -13814,6 +13907,98 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
+          1. If _baseValue_ is *undefined* or *null*, then
+            1. Return *undefined*.
+          1. Let _optionalChain_ be |OptionalChain|.
+          1. Return ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
```
```suggestion
          1. Return the result of performing ChainEvaluation of _optionalChain_ with arguments _baseValue_ and _baseReference_.
```

bakkot commented on spec.html:
```diff
@@ -13814,6 +13907,98 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
+            MemberExpression OptionalChain
+        </emu-grammar>
+        <emu-alg>
+          1. Let _baseExpression_ be |MemberExpression|.
```
I'm not sure the `_baseExpression_` and `_optionalChain_` aliases add any clarity in these algorithms. Would it not be better to just have

```
1. Let _baseReference_ be the result of evaluating |MemberExpression|.
```
and similarly
```
1. Return the result of performing ChainEvaluation of |OptionalChain| with arguments _baseValue_ and _baseReference_.
```
thereby avoiding a layer of indirection? That's what the rest of the spec tends to do.

claudepache replied:
> I'm not sure the _baseExpression_ and _optionalChain_ aliases add any clarity in these algorithms. 

It is an undue leftover of the original spec, [that said](https://tc39.es/proposal-optional-chaining/#sec-optional-chaining-evaluation): “Let _baseExpression_ be the first child of this _OptionalExpression_ (i.e., _MemberExpression_, _CallExpression_, or _OptionalExpression_)”. Now that the algorithm is repeated thrice (once per possible first child of _OptionalExpression_), there is indeed no point in having _baseExpression_.

bakkot commented on spec.html:
```diff
@@ -13814,6 +13907,98 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
+        <emu-grammar>OptionalChain : OptionalChain `[` Expression `]`</emu-grammar>
+        <emu-alg>
+          1. Let _optionalChain_ be |OptionalChain|.
+          1. Let _newReference_ be ? ChainEvaluation of _optionalChain_ with arguments _baseValue_ and _baseReference_).
```
```suggestion
          1. Let _newReference_ be ? ChainEvaluation of _optionalChain_ with arguments _baseValue_ and _baseReference_.
```

bakkot commented on spec.html:
```diff
@@ -13428,6 +13479,19 @@ <h1>Static Semantics: Contains</h1>
+        <emu-grammar>OptionalChain : `?.` IdentifierName</emu-grammar>
+        <emu-alg>
+          1. If _symbol_ is a |ReservedWord|, return *false*.
+          1. If _symbol_ is an |Identifier| and StringValue of _symbol_ is the same value as the StringValue of |IdentifierName|, return *true*.
```
If https://github.com/tc39/ecma262/pull/1519 lands before this PR, this step will need to be removed both places it occurs.

bakkot commented on spec.html:
```diff
@@ -13584,29 +13654,52 @@ <h1>Runtime Semantics: Evaluation</h1>
+      <h1>Runtime Semantics: EvaluateIdentifierKeyPropertyAccess( _baseValue_, _identifierName_, _strict_ )</h1>
+      <p>The abstract operation EvaluateIdentifierKeyPropertyAccess takes as arguments a value _baseValue_, a Parse Node _identifierName_, and a Boolean argument _strict_. It performs the following steps:</p>
+      <emu-alg>
+        1. Let _bv_ be ? RequireObjectCoercible(_baseValue_).
```
I would add
```
1. Assert: _identifierName_ is a |IdentifierName|
```
before this line, to justify the use of StringValue below.

rkirsling commented:
Hmm, `Expression` / `Identifier` is one thing, but I don't think I understand why `Key` is in there.

DanielRosenwasser commented:
@rkirsling it was just to disambiguate that the *Identifier* or *Expression* is the key, not the base value.

rkirsling commented:
Not sure how the others feel, but that suggestion sounds pretty good to me.

michaelficarra commented:
I really hate bikeshedding something as unimportant as this, but really, "Identifier"/"IdentifierName" should not be in these names.

syg commented:
I am fine with literally anything for those names so long as it doesn't require me to intuit what's dynamic and what's static.

devsnek commented:
In engine262 we'd call them `EvaluateMemberExpressionWithIdentifier` and `EvaluateMemberExpressionWithExpression`

DanielRosenwasser commented:
I think using the syntactically-guided names (as it currently stands) is probably the best way forward. I'm also fine with switching to `EvaluatePropertyAccessWith[Identifier|Expression]Key` if there's a preference there.

Keep in mind that this can be switched later if it causes legitimate issues, so I'd urge people not to keep picking out new colors for the bikeshed. 😄

rkirsling commented:
I'm fine with virtually everything that's been suggested (particularly the recent suggestions involving `By` / `With`), but I do think the current naming in particular is hard to parse. It makes me ask, "What's a _key property_? Or is it a _key property access_?"

rkirsling commented on spec.html:
```diff
@@ -21710,6 +21890,34 @@ <h1>Expression Rules</h1>
+        <emu-alg>
+          1. Return *false*.
+        </emu-alg>
+        <emu-grammar>
```
A final nitpick: Given that the [ordering of these productions was changed](https://github.com/tc39/ecma262/pull/1646#discussion_r345970102), we should probably put the Arguments case before the Expression and IdentifierName cases here too.

rkirsling replied:
(Then again, I guess the Member-Call-Optional ordering isn't being preserved here either...hmm.)
